### PR TITLE
Add section on hyperparameter tuning

### LINKED ISSUE #351: Add hyperparameter tuning to episode 3
Add explanation of hyperparamter tuning to episode 3

In this section: https://carpentries-incubator.github.io/deep-learning-intro/03-monitor-the-model/index.html#counteract-model-overfitting we are actually doing hyperparamter tuning (we are changing a hyperparameter). It is only a small extra step to do this within a for-loop. 

- [ ] ~~Add an infobox that discusses hyperparameter tuning in a for-loop~~ see https://github.com/carpentries-incubator/deep-learning-intro/issues/351#issuecomment-1621589631
- [ ] Have a callout to explain grid search.

You can get inspiration from @qualiaMachine 's script here: https://colab.research.google.com/drive/1B_F5ik6P9hfbKT2pzfMgsvFeF99_Uy9V?usp=sharing

This issue resulted from discussions in #348

### PR DESCRIPTION
**Pull request by svenvanderburg**
Fixes #351 

@reviewer: This is now in episode 4, I was thinking to maybe put it together with transfer learning in the new episode 5 (See #349). Episode 5 could then be called 'advanced deep learning concepts'.

### CODE REVIEW

File: episodes/4-advanced-layer-types.Rmd [hunk 1]
```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
...
+    inputs = keras.Input(shape=train_images.shape[1:])
+    x = inputs
+    for layer in range(n_layers):
+        x = keras.layers.Conv2D(50, (3, 3), activation='relu')(x)
+        x = keras.layers.MaxPooling2D((2, 2))(x)
+    x = keras.layers.Dropout(dropout_rate)(x)
+    x = keras.layers.Flatten()(x)
+    x = keras.layers.Dense(50, activation='relu')(x)
+    outputs = keras.layers.Dense(10)(x)
+    model = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
```
Comment (carschno) [inline episodes/4-advanced-layer-types.Rmd:812]: Just a reminder for later: this (and all other references to `cifar_model`) needs to be adapted to the changes in #445 .

File: episodes/4-advanced-layer-types.Rmd [hunk 2]
```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
...
+    return model
+```
+
+Now, let's find the best combination of hyperparameters using grid search. 
+Grid search is the simplest hyperparameter tuning strategy, 
+you just test all the combinations of a few predefined values for the hyperparameters that you want to vary.
+Note that this can take a some time to train (around 5 minutes or longer).
```
Comment (carschno) [inline episodes/4-advanced-layer-types.Rmd:819]:
```suggestion
Note that this can take some time to train (around 5 minutes or longer).
```

File: episodes/4-advanced-layer-types.Rmd [hunk 3]
```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
...
+- A. 6 different models were trained in this grid search run, because there are 6 possible combinations for the defined hyperparameter values
+- B. There are 2 different models trained, 1 for each hyperparameter that we want to change
+- C. 1 model is trained with 6 different hyperparameter combinations
+- D. The model with 1 layer and a dropout rate of 0.2 is the best model with a validation loss of 1.45
+- E. The model with 2 layers and a dropout rate of 0.5 is the best model with a validation loss of 1.24
+- F. We have found the model with the best possible combination of dropout rate and number of layers
+
+2 (Optional): Perform a grid search finding the best combination of the following hyperparameters: 2 different activation functions: 'relu', and 'tanh', and 2 different values for the kernel size: 3 and 4. Which combination works best?
+
+**Hint**: You can use `hp.Choice("name", ["value1", "value2"])` to use hyperparameters from a predefined set of possible values.
```
Comment (carschno) [inline episodes/4-advanced-layer-types.Rmd:909]: This hint seems a bit confusing without further explanation. Given that this is an optional extra exercise, it might be ok to leave it to the learner.

File: episodes/4-advanced-layer-types.Rmd [hunk 4]
```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
...
+```
+A kernel size of 3 and `relu` as activation function is the best tested combination.
+
+::::
+::: 
+
+Grid search can quickly result in a combinatorial explosion because all combinations of hyperparameters are trained and tested.
+Instead, `random search` randomly samples combinations of hyperparemeters, allowing for a much larger look through a large number of possible hyperparameter combinations.
+
+Next to grid search and random search there are many different hyperparameter tuning strategies, including `neural architecture search` where a separate neural network is trained to find the best architecture for a model!
```
Comment (carschno) [inline episodes/4-advanced-layer-types.Rmd:1002]: Maybe add a link (to Wikipedia?) for more information about neural architecture search?

```suggestion
Next to grid search and random search there are many different hyperparameter tuning strategies, including [neural architecture search](https://en.wikipedia.org/wiki/Neural_architecture_search) where a separate neural network is trained to find the best architecture for a model!
```

**Review by carschno — APPROVED** [review]
Looks good to me. I added a few minor remarks.
This PR is unrelated to the changes here, but perhaps a good occasion to merge in: https://github.com/carpentries-incubator/deep-learning-intro/pull/451

File: episodes/4-advanced-layer-types.Rmd (hunk 1 above)
Comment (svenvanderburg) [reply episodes/4-advanced-layer-types.Rmd:812]: Yes, this is done in #448

File: episodes/4-advanced-layer-types.Rmd (hunk 3 above)
Comment (svenvanderburg) [reply episodes/4-advanced-layer-types.Rmd]: Fixed in [c38b110](https://github.com/carpentries-incubator/deep-learning-intro/pull/437/commits/c38b110c0e25e16c087c1038278608e7d6617cf1)

File: episodes/4-advanced-layer-types.Rmd [hunk 5]
```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
...
 :::
 
+## Hyperparameter tuning
+::: instructor
+## Do a live demo instead of live coding
+You might want to demonstrate this section on hyperparamater tuning instead of doing live coding.
+The goal is to show that hyperparameter tuning can be done easily with `keras_tuner`, not to memorize all the exact syntax of how to do it. This will probably save you half an hour of participants typing over code that they already know from before. In addition, on really slow machines running the grid search could possibly take more than 10 minutes.
+:::
+
+In general if you are varying hyperparameters (such as the dropout rate) to find the combination of hyperparameters with the best model performance this is called hyperparameter tuning. A naive way to do this is to write a for-loop and train a slightly different model in every cycle.
```
[the remark this answers is not in the document]
Comment (svenvanderburg) [reply episodes/4-advanced-layer-types.Rmd]: Fixed in [c38b110](https://github.com/carpentries-incubator/deep-learning-intro/pull/437/commits/c38b110c0e25e16c087c1038278608e7d6617cf1)

**Comment by svenvanderburg** [comment]
Thanks for reviewing @carschno. We discussed that this will conflict with #448, in the sense that the output and solutions of the exercises will not match the new dataset. To be updated!

**Comment by svenvanderburg** [comment]
I made the results reflect the new dollarstreet dataset, since I didn't change any content since @carschno' review, I will merge this now.
